cbrain_task/bids_app_handler/bourreau/bids_app_handler.rb

```diff
@@ -0,0 +1,226 @@
+    mode         = params[:_cb_mode]
+
+    # Check that we have a zero status code from bosh
+    if mode != 'save' # a 'save' task doesn't produce a status file
+      status = File.read("status.all/#{run_id}").strip rescue 'No status file'
+      if status != '0'
+        self.addlog("Bosh did not return a success status. Got: #{status}")
+        return false
+      end
+    end
+
+    # Check that we have at least one output file per participant
+    outputdir            = output_dir_basename_for_batch()
+    glob                 = Dir.glob "#{outputdir}/*"
+    all_ok               = true
+    participants_outputs = {}
+    selected_participants.each do |sub|
+      sublist = glob.select { |p| Pathname.new(p).basename.to_s =~ /^sub-#{Regexp.quote(sub)}(\b|_)/ }
+      participants_outputs[sub] = sublist
+      if sublist.present?
+        self.addlog "Found #{sublist.size} output files for participant '#{sub}'"
+      else
+        self.addlog "Error: can't find outputs for participant '#{sub}'"
+        all_ok = false
+      end
+    end
+
+    # We don't do any saving for participants tasks
+    if mode == 'participant'
+      self.addlog("No output files need saving here, a separate task handles that.")
+      return all_ok
+    end
+
+    # For provevance logs at the end
+    created_outputs = []
+
+    # For mode 'save', we just save each participant's result
+    if mode == 'save'
+      selected_participants.each do |sub|
+        sublist = participants_outputs[sub]
+        sublist.each do |one_output| # "outdir/something"
+          out_type = File.file?(one_output) ? SingleFile : FileCollection
+          basename = Pathname.new(one_output).basename.to_s #hopefully respects CBRAIN legal characters
+          cb_out = safe_userfile_find_or_new(out_type, { :name => basename + "_" + run_id() })
+          cb_out.cache_copy_from_local_file(one_output)
+          cb_out.move_to_child_of( bids_dataset )
+          created_outputs << cb_out
+        end
+      end
+    end
+
+    # For mode 'group', we just save the output dir containing all participants outputs
+    if mode == 'group'
+      group_output_name = 'group_output' + "_" + run_id()
+      self.addlog("Attempting to save group results '#{group_output_name}'")
+      cb_out = safe_userfile_find_or_new(FileCollection, { :name => group_output_name })
+      cb_out.cache_copy_from_local_file(outputdir)
+      cb_out.move_to_child_of(bids_dataset)
+      created_outputs << cb_out
+    end
+
+    # Add provenance logs
+    self.addlog_to_userfiles_these_created_these( bids_dataset, created_outputs )
+
+    return true
+  end
+
+  protected
+
+  def output_dir_basename_for_batch #:nodoc:
+    batch_id = self.batch_id || self.id
+    "outdir-#{batch_id}-#{run_number}"
+  end
+
+  def bosh_is_supported? #:nodoc:
+    out, err = self.tool_config_system("bosh version")
+
+    if err.present?
+      self.addlog("No proper 'bosh' program detected on system. Got stderr:")
```

Comment:
Recommend we say "Boutiques" rather than "bosh" to be clear that it isn't that stupid python BOSH that we all accidentally installed.

Reply:
Hum, OK.

Reply:
Done.